com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java

```diff
@@ -0,0 +1,240 @@
+     * @param input
+     *              the input stream
+     * @param output
+     *              the output stream
+     */
+    public AbstractProtocolServer(InputStream input, OutputStream output) {
+        this.reader = new BufferedReader(new InputStreamReader(input, PROTOCOL_ENCODING));
+        this.writer = new PrintWriter(new BufferedWriter(new OutputStreamWriter(output, PROTOCOL_ENCODING)));
+        this.contentLength = -1;
+        this.rawData = new ByteBuffer();
+        this.eventQueue = new ConcurrentLinkedQueue<>();
+    }
+
+    /**
+     * A while-loop to parse input data and send output data constantly.
+     */
+    public void start() {
+        char[] buffer = new char[BUFFER_SIZE];
+        try {
+            while (!this.terminateSession) {
+                int read = this.reader.read(buffer, 0, BUFFER_SIZE);
+                if (read == -1) {
+                    break;
+                }
+
+                this.rawData.append(new String(buffer, 0, read).getBytes(PROTOCOL_ENCODING));
+                this.processData();
+            }
+        } catch (IOException e) {
+            logger.log(Level.SEVERE, String.format("Read data from io exception: %s", e.toString()), e);
+        }
+    }
+
+    /**
+     * Sets terminateSession flag to true. And the dispatcher loop will be terminated after current dispatching operation finishes.
+     */
+    public void stop() {
+        this.terminateSession = true;
+    }
+
+    public boolean isTerminating() {
+        return this.terminateSession;
+    }
+
+    /**
+     * Send event to DA immediately.
+     * @param eventType
+     *                 event type
+     * @param body
+     *                 event body
+     */
+    protected void sendEvent(String eventType, Object body) {
+        sendMessage(new Messages.Event(eventType, body));
+    }
+
+    /**
+     * If the the dispatcher is idle, then send the event immediately.
+     * Else add the new event to an eventQueue first and send them when dispatcher becomes idle again.
+     * @param eventType
+     *              event type
+     * @param body
+     *              event content
+     */
+    protected void sendEventLater(String eventType, Object body) {
+        synchronized (this) {
+            if (this.isDispatchingData) {
+                this.eventQueue.offer(new Messages.Event(eventType, body));
+            } else {
+                sendMessage(new Messages.Event(eventType, body));
+            }
+        }
+    }
+
+    protected void sendMessage(Messages.ProtocolMessage message) {
+        message.seq = this.sequenceNumber.getAndIncrement();
+
+        String jsonMessage = JsonUtils.toJson(message);
+        byte[] jsonBytes = jsonMessage.getBytes(PROTOCOL_ENCODING);
+
+        String header = String.format("Content-Length: %d%s", jsonBytes.length, TWO_CRLF);
+        byte[] headerBytes = header.getBytes(PROTOCOL_ENCODING);
+
+        byte[] data = new byte[headerBytes.length + jsonBytes.length];
```

Comment:
Sounds some duplication? Should be ByteBuffer?

Reply:
Nice catch. fixed.